src/sinks/datadog_archives.rs

```diff
@@ -0,0 +1,585 @@
+#[serde(deny_unknown_fields)]
+pub struct S3Config {
+    #[serde(flatten)]
+    pub options: S3Options,
+    #[serde(flatten)]
+    pub region: RegionOrEndpoint,
+    #[serde(default)]
+    pub auth: AwsAuthentication,
+}
+
+#[derive(Clone, Debug, Default, Deserialize, Serialize)]
+#[serde(deny_unknown_fields)]
+pub struct S3Options {
+    acl: Option<S3CannedAcl>,
+    grant_full_control: Option<String>,
+    grant_read: Option<String>,
+    grant_read_acp: Option<String>,
+    grant_write_acp: Option<String>,
+    server_side_encryption: Option<S3ServerSideEncryption>,
+    ssekms_key_id: Option<String>,
+    storage_class: Option<S3StorageClass>,
+    tags: Option<BTreeMap<String, String>>,
+}
+
+impl GenerateConfig for DatadogArchivesSinkConfig {
+    fn generate_config() -> toml::Value {
+        toml::Value::try_from(Self {
+            service: "".to_owned(),
+            bucket: "".to_owned(),
+            key_prefix: None,
+            request: TowerRequestConfig::default(),
+            aws_s3: Some(S3Config::default()),
+        })
+        .unwrap()
+    }
+}
+
+#[derive(Debug, Snafu, PartialEq)]
+enum ConfigError {
+    #[snafu(display("Unsupported service: {:?}", service))]
+    UnsupportedService { service: String },
+    #[snafu(display("Unsupported storage class: {:?}", storage_class))]
+    UnsupportedStorageClass { storage_class: String },
+}
+
+const KEY_TEMPLATE: &'static str = "/dt=%Y%m%d/hour=%H/";
+
+impl DatadogArchivesSinkConfig {
+    pub fn new(&self, cx: SinkContext) -> crate::Result<(super::VectorSink, super::Healthcheck)> {
+        let request = self.request.unwrap_with(&TowerRequestConfig {
+            concurrency: Concurrency::Fixed(50),
+            rate_limit_num: Some(250),
+            ..Default::default()
+        });
+
+        let bucket = self.bucket.clone();
+        let prefix = self.key_prefix.clone();
+
+        let (svc, healthcheck) = match &self.service[..] {
+            "aws_s3" => {
+                let s3_config = self.aws_s3.as_ref().expect("s3 config wasn't provided");
+                let client =
+                    aws_s3::create_client(&s3_config.region, &s3_config.auth, None, &cx.proxy)?;
+                let svc = self
+                    .build_s3_service(request, bucket.clone(), prefix, client.clone())
+                    .map_err(|error| format!("{:?}", error))?;
+                Ok((svc, aws_s3::healthcheck(bucket, client).boxed()))
+            }
+
+            service => Err(ConfigError::UnsupportedService {
+                service: service.to_owned(),
+            }),
+        }?;
+
+        // We should avoid producing many small batches, therefore we use settings recommended by DD:
+        // batch size - 100mb
+        // batch timeout - 15min
+        let batch = BatchSettings::default().bytes(100_000_000).timeout(900);
+        let buffer = PartitionBuffer::new(Buffer::new(batch.size, Compression::gzip_default()));
```

Comment:
it's not really documented anywhere unfortunately, it's just settings used by DD Logs internally for a similar service. The idea behind this is to avoid producing small batches which might slow down rehydration jobs.